sd_meh/merge_methods.py

```diff
@@ -209,3 +211,76 @@ def filter_top_k(a: Tensor, k: float):
     k_value, _ = torch.kthvalue(torch.abs(a.flatten()).float(), k)
     top_k_filter = (torch.abs(a) >= k_value).float()
     return a * top_k_filter
+
+
+def rotate(a: Tensor, b: Tensor, alpha: float, beta: float, **kwargs):
+    if alpha == 0 and beta == 0:
+        return a
+
+    is_conv = len(a.shape) == 4 and a.shape[-1] != 1
+    if len(a.shape) == 0 or is_conv or torch.allclose(a, b):
+        return weighted_sum(a, b, beta)
+
+    if len(a.shape) == 4:
+        shape_2d = (-1, functools.reduce(operator.mul, a.shape[1:]))
+    else:
+        shape_2d = (-1, a.shape[-1])
+
+    a_neurons = a.reshape(*shape_2d).double()
+    b_neurons = b.reshape(*shape_2d).double()
+
+    a_centroid = a_neurons.mean(0)
+    b_centroid = b_neurons.mean(0)
+    new_centroid = sample_ellipsis(a_centroid, b_centroid, 2 * torch.pi * alpha)
+    a_neurons -= a_centroid
+    b_neurons -= b_centroid
+
+    if len(a.shape) == 1 or len(a.shape) == 2 and a.shape[0] == 1:
+        return new_centroid.reshape_as(a)
+
+    svd_driver = "gesvd" if a.is_cuda else None
```

Comment:
This is actually a lot more complex than meets the eye. We should be determining driver based on the size of the matrix. Different drivers perform faster on smaller/bigger matrices. And in some instances the CPU will out perform the GPU. What exactly is our average matrix size when we call svd?

Reply:
If we include all keys, it goes form $320^2$ to $~20K^2$. if we exclude all conv layers (which have the largest neurons), the upper bound is $~5K^2$. I can list all sizes in a bit, they all are square matrices.

Reply:
all matrices sizes that goes through the svd are listed below:
- 320x320
- 640x640
- 768x768
- 1280x1280
- 2560x2560
- 5120x5120

Reply:
I did some benchmarking between jax's svd functions jitted through XLA and pytorch's different drivers on a colab using a v100 (a 3080 is about equal to this in PyTorch Performance), and these were the results.
![image](https://github.com/s1dlx/meh/assets/77745122/af8e2946-f5d3-44c4-b72a-92342c13f712)
Basically unless you need full accuracy, even with full_matrices set to true, gesvdj is going to be faster. However the speed you gain comes at the cost of some accuracy, and the potential to not always converge without needing to fall back to gesvd.

Reply:
I tested gesvdj with fractional alpha and the error builds up pretty quickly. IIRC it works okay in the non fractional cases (alpha = 0 or alpha = 1). I'll run some tests again to verify this and make sure this was the cause.

Reply:
Well I can't even merge using gesvdj, on my 3080 it takes 1m10s to merge a single 3072x3072 key. The full merge might take more than a hour.

Reply:
Yeah, something seems wrong on your end, are you sure that PyTorch is seeing your GPU?

Reply:
Unless of course I did not switch the driver correctly. All I did was use the string `"gesvdj"` instead of `"gesvd"`.

Reply:
I'll try updating pytorch to 2.1.x and see if that fixes it.

Reply:
Updating did not increase merge speed with gesvdj.

Reply:
How long does it take on your end to merge the same set of models with gesvd and then with gesvdj?

Reply:
How long does it take on your end to merge the same set of models with gesvd and then with gesvdj? I understand that the benchmarks above indicate that gesvdj should be faster, but I wonder if we should run benchmarks against the particular type of data we wound see with model merging in case gesvdj runs faster or slower depending on what the value of the input is.

Reply:
I'm getting 1.3 sec/it to 1.8it/sec with gesvd. Let me see what gesvdj looks like. gesvdj does seem slower. My guess is that our matrices aren't converging. Now, do we need the full matrix from svd? Setting that to false would help speed things up as well.

Reply:
There is no apparent quality loss, so I'll commit `full_matrices=False`.

Reply:
If we have a lot of similar matrices, it might be worth it to try jax instead of pytorch and use the XLA accelerated svd functions there. From what I understanding, the caching will make repeated runs of similar data faster.

Reply:
Also, I am currently having an issue where moving beyond stage 1 causes a crash with the following backtrace on cpu.

```
stage 1:  98%|██████████████████████████████████████████████████████████▌ | 1105/1131 [04:51<00:06,  3.79it/s]
Using already loaded model bbwm_Aurora.safetensors [431eebe43e]: done in 8.9s (send model to device: 8.9s)
*** API error: POST: http://127.0.0.1:7860/bbwm/merge-models {'error': 'RuntimeError', 'detail': '', 'body': '', 'errors': 'torch.linalg.eig: input tensor should not contain infs or NaNs.'}
    Traceback (most recent call last):
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/anyio/streams/memory.py", line 98, in receive
        return self.receive_nowait()
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/anyio/streams/memory.py", line 93, in receive_nowait
        raise WouldBlock
    anyio.WouldBlock

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 78, in call_next
        message = await recv_stream.receive()
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/anyio/streams/memory.py", line 118, in receive
        raise EndOfStream
    anyio.EndOfStream

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/home/maria/stable-diffusion/stable-diffusion-webui/modules/api/api.py", line 187, in exception_handling
        return await call_next(request)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 84, in call_next
        raise app_exc
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 70, in coro
        await self.app(scope, receive_or_disconnect, send_no_error)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 108, in __call__
        response = await self.dispatch_func(request, call_next)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/modules/api/api.py", line 151, in log_and_time
        res: Response = await call_next(req)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 84, in call_next
        raise app_exc
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/base.py", line 70, in coro
        await self.app(scope, receive_or_disconnect, send_no_error)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/cors.py", line 84, in __call__
        await self.app(scope, receive, send)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 24, in __call__
        await responder(scope, receive, send)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/gzip.py", line 44, in __call__
        await self.app(scope, receive, self.send_with_gzip)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
        raise exc
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
        await self.app(scope, receive, sender)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
        raise e
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
        await self.app(scope, receive, send)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/routing.py", line 718, in __call__
        await route.handle(scope, receive, send)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/routing.py", line 276, in handle
        await self.app(scope, receive, send)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/starlette/routing.py", line 66, in app
        response = await func(request)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/fastapi/routing.py", line 237, in app
        raw_response = await run_endpoint_function(
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/fastapi/routing.py", line 163, in run_endpoint_function
        return await dependant.call(**values)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/extensions/sd-webui-bayesian-merger/scripts/api.py", line 78, in merge_models_api
        merged = merge_models(
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/sd_meh/merge.py", line 162, in merge_models
        merged = simple_merge(
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/sd_meh/merge.py", line 245, in simple_merge
        res.result()
      File "/usr/lib/python3.10/concurrent/futures/_base.py", line 451, in result
        return self.__get_result()
      File "/usr/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
        raise self._exception
      File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
        result = self.fn(*self.args, **self.kwargs)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/sd_meh/merge.py", line 353, in simple_merge_key
        with merge_key_context(key, thetas, *args, **kwargs) as result:
      File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
        return next(self.gen)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/sd_meh/merge.py", line 450, in merge_key_context
        result = merge_key(*args, **kwargs)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/sd_meh/merge.py", line 422, in merge_key
        merged_key = merge_method(**merge_args).to(device)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/sd_meh/merge_methods.py", line 252, in rotate
        transform = fractional_matrix_power(transform, alpha)
      File "/home/maria/stable-diffusion/stable-diffusion-webui/venv/lib/python3.10/site-packages/sd_meh/merge_methods.py", line 272, in fractional_matrix_power
        eigenvalues, eigenvectors = torch.linalg.eig(matrix)
    RuntimeError: torch.linalg.eig: input tensor should not contain infs or NaNs.

---
```

Reply:
> on CPU [...] the performance is significantly better

5 minutes is the time it takes on my system to merge on cuda with the gesvd driver. I'll see if I get a speed discount on cpu with gesvdj.

> Also, I am currently having an issue where moving beyond stage 1 causes a crash with the following backtrace on cpu.

Although I am not running this through the bbmw API, it is probably caused by the merge method. It seems to be near the end of stage 1 (still in it because the exception is raised inside the merge method). It seems to be caused by imprecision in the svd calculation.

Reply:
Reading from the docs, it doesn't seem to be possible to pick a specific driver if the device is not cuda:
> The `driver` kwarg may be used in CUDA with a cuSOLVER backend to choose the algorithm used to compute the SVD. The choice of a driver is a trade-off between accuracy and speed.

Reply:
Sometimes `torch.linalg.cond(a_neurons)` is very large (~1.0e+15) and at other times it is < 1000. According to the docs:

> - If `A` is well-conditioned (its [condition number](https://pytorch.org/docs/master/linalg.html#torch.linalg.cond) is not too large), or you do not mind some precision loss.
>    - For a general matrix: ‘gesvdj’ (Jacobi method)
>    - If A is tall or wide (m >> n or m << n): ‘gesvda’ (Approximate method)

So `gesvda` is definitely not appropriate, and gesvdj seems to be sometimes (but most keys seem to have a large cond experimentally). Calculating `torch.linalg.cond` takes a bit of time, so it may or may not be worth it to switch to gesvdj conditionally.

Reply:
Do you know how I can install pytorch_xla? I'm not able to run the pip install here:
https://github.com/pytorch/xla/#getting-started

> Looking in links: https://storage.googleapis.com/libtpu-releases/index.html
> Requirement already satisfied: torch~=2.1.0 in d:\src\meh-dev\venv\lib\site-packages (2.1.1+cu118)
> ERROR: Could not find a version that satisfies the requirement torch_xla[tpu]~=2.1.0 (from versions: 1.0)
> ERROR: No matching distribution found for torch_xla[tpu]~=2.1.0

Reply:
The quickest run I could get was under 5 minutes (~4m45s) using `--work_device cuda` and `--threads 10`. IIUC threads are useful here to reduce the overhead of tensors being copied to another device.

Reply:
So I actually did a test with a full matrices turned on and the issue still occurred. So I'm not sure what is causing it.

Reply:
> Do you know how I can install pytorch_xla? I'm not able to run the pip install here: https://github.com/pytorch/xla/#getting-started
> 
> > ```
> > Looking in links: https://storage.googleapis.com/libtpu-releases/index.html
> > Requirement already satisfied: torch~=2.1.0 in d:\src\meh-dev\venv\lib\site-packages (2.1.1+cu118)
> > ERROR: Could not find a version that satisfies the requirement torch_xla[tpu]~=2.1.0 (from versions: 1.0)
> > ERROR: No matching distribution found for torch_xla[tpu]~=2.1.0
> > ```

So when I referenced XLA, I was using it in the context of JAX, where you precompile your function using the XLA JIT with JAX, and then go from there. You can interoperate PyTorch and JAX using DLpacks so there shouldn't be any issues with type compatibility, the notation might get a little messy though.

Reply:
By the way `full_matrices=False` doesn't produce a reduced SVD when $m=n$ ($m$ and $n$ being the width and height of the svd input). That's why it didn't seem to affect generation speed. We might want to remove it as it doesn't really change anything, since the input to the svd is always a square covariance matrix here.

Reply:
So I did complete a full merge on CUDA, and didn't receive the error. I think it has something to do with trying to move models between the CPU and GPU, interacting with WebUI keeping models loaded in memory. Is there sanity checking when the models are loaded to ensure that they have been moved to CPU if the work_device is set to CPU?

Reply:
Before merging, when assembling the merge args, the weihgts are sent to the requested device:
https://github.com/s1dlx/meh/blob/2780321723efe914aac876bdb8cbf203d1dbee1c/sd_meh/merge.py#L465-L466
note that if `work_device` is `None`, it takes the value of `device`:
https://github.com/s1dlx/meh/blob/2780321723efe914aac876bdb8cbf203d1dbee1c/sd_meh/merge.py#L371-L372

Reply:
I think I found the culprit.

It seems that on CPU there isn't enough precision sometimes, which leads too $U$ or $V^T$ having a determinant of 0. This is not what SVD should output, $U$ and $V^T$ should always be orthogonal transforms, which implies $|det U| = |det V^T| = 1$.

When the determinant of $U$ or $V^T$ is 0, then this line divides by 0:
https://github.com/s1dlx/meh/blob/f506831822e780d039c921e85975229db9f4a19a/sd_meh/merge_methods.py#L248
```python
        u[:, -1] /= torch.det(u) * torch.det(v_t)
```

Reply:
As noted below, while this prevents the entire merge from raising an error, rotations with invalid determinants still result in a broken merge. I went the other direction and raised an error instead.